### DISCUSSION TITLE
Is there any way to write from one client and read from other client via blobfuse2?

### DISCUSSION THREAD

**Discussion by jejass** [Azure/azure-storage-fuse#1552] [Q&A]
Hello team,

Here is the scenario. Both client A and client B mount Storage account via blobfuse2. Client A is responsible for write, Client B is responsible for read.

I test 2 configurations, with both of them, client B cannot read the latest updated blob, even after waiting for 1 hour.

**configuration 1**:

logging:
  type: syslog
  level: log_debug

components:
  - libfuse
  - file_cache
  - attr_cache
  - azstorage

libfuse:
  attribute-expiration-sec: 0
  entry-expiration-sec: 0
  negative-entry-expiration-sec: 0

**file_cache:
  path: /mnt/blobfusetmp
  timeout-sec: 0
  max-size-mb: 0
  refresh-sec: 1**

attr_cache:
  timeout-sec: 0

**configuration 2:**
logging:
  type: syslog
  level: log_debug

components:
  - libfuse
  - file_cache
  - attr_cache
  - azstorage

libfuse:
  attribute-expiration-sec: 0
  entry-expiration-sec: 0
  negative-entry-expiration-sec: 0

**block_cache:
  block-size-mb: 1
  mem-size-mb: 1000
  path: /mnt/blobfusetmp
  disk-size-mb: 1
  disk-timeout-sec: 1**

attr_cache:
  timeout-sec: 0

I understand 'direct_io' parameter can disable kernal page cache. but with this parameter, the application will report an error.

So, I want to figure out a way in the blobfuse2 configuration to make Client B can read the latest update.

**is that possible?** 

as per my testing, it seems the Client B will read the blob once. After then, the read Client B will not refresh the content, even with timeout 0, cache size 0. I found the parameter 'refresh-sec':  <number of seconds after which compare lmt of file in local cache and container and refresh file if container has the latest copy>. but it does not work as well.

**is that how blobfuse2 works? the read client will never refresh the content in any configuration?**

thank you in advance.

**Comment by souravgupta-msft** [comment]
Hi @jejass. Thanks for reaching out. You are not seeing updated data in Client B because of kernel caching. To disable kernel cache, you will have to use `direct_io` or manually clear the kernel cache using the below command before read operations,
`echo 2 > /proc/sys/vm/drop_caches`

**Reply by jejass** [reply]
@souravgupta-msft ,thanks for your response. the thing is here, cx is not able to use direct_io as their application will report error message when using direct_io. It looks like something wrong due to the C++ shared memory model.

is there any way from blobfuse side to make the kernal cache expiry then reload the data from Storage account?

**Reply by souravgupta-msft** [reply]
Can you try with this parameter, `--disable-writeback-cache=true`
